### PR TITLE
Add docs on building your own service graphs

### LINKED ISSUE #3341: [DOC] How to build your own service graphs using the Node Graph visualisation from Tempo generated metrics
Customers have asked for it in the past, we should really document the queries and a couple of examples of how to build this out in a dashboard panel.

### PR DESCRIPTION
**Pull request by ilyamochalov**
**What this PR does**: Adds docs on building your own service graphs

**Which issue(s) this PR fixes**:
Fixes #3341 

**Checklist**
- [ ] Tests updated
- [x] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Review by joe-elliott — APPROVED** [review]
lgtm

pinging @knylander-grafana, our primary docs maintainer, who may have some stylistic or organizational thoughts

thanks for the contribution!

File: docs/sources/tempo/metrics-generator/service_graphs/custom-service-graphs.md
Comment (knylander-grafana) [inline docs/sources/tempo/metrics-generator/service_graphs/custom-service-graphs.md:4]: Rewording slightly so we don't have a compound noun using the product name. 
```suggestion
description: Build your own service graphs using the Node Graph visualization from metrics generated by Tempo. 
```

File: docs/sources/tempo/metrics-generator/service_graphs/custom-service-graphs.md
Comment (knylander-grafana) [inline docs/sources/tempo/metrics-generator/service_graphs/custom-service-graphs.md:10]:
```suggestion
You can craft custom service graphs based on the metrics generated from metrics-generator.  
```

File: docs/sources/tempo/metrics-generator/service_graphs/custom-service-graphs.md
Comment (knylander-grafana) [inline docs/sources/tempo/metrics-generator/service_graphs/custom-service-graphs.md:15]: Great that you have a prereq section! Our other doc pages use "Before you begin," with the same kind of content. I've also changed the heading level up one and removed "Building custom graphs", since the content for building custom graphs really starts in the section section. 
```suggestion
## Before you begin  
```

File: docs/sources/tempo/metrics-generator/service_graphs/custom-service-graphs.md
Comment (knylander-grafana) [inline docs/sources/tempo/metrics-generator/service_graphs/custom-service-graphs.md:24]:
```suggestion
For custom service graphs, you'll use:
- Tempo metrics-generator
- Grafana deployment of version v10 or higher 
- Grafana [Node graph panel](https://grafana.com/docs/grafana/latest/panels-visualizations/visualizations/node-graph/)

The metrics-generator creates multiple metrics, including `traces_service_graph_request_total`. This metric holds:
- relationships between services
- total number of requests performed between services
```

File: docs/sources/tempo/metrics-generator/service_graphs/custom-service-graphs.md
Comment (knylander-grafana) [inline docs/sources/tempo/metrics-generator/service_graphs/custom-service-graphs.md:35]:
```suggestion
### Create one or more Grafana dashboards

1. In Grafana, [create a new dashboard](https://grafana.com/docs/grafana/latest/dashboards/build-dashboards/create-dashboard/).
1.  Add two variables:
  - data source of type `Prometheus`
  ![Custom service graph Grafana data source variable](../custom-service-graph-var-datasource.png)
  - service of type `Label values`, enable multi-value option
  ![Custom service graph service variable](../custom-service-graph-var-service.png)

### Add a panel

```

File: docs/sources/tempo/metrics-generator/service_graphs/custom-service-graphs.md
Comment (knylander-grafana) [inline docs/sources/tempo/metrics-generator/service_graphs/custom-service-graphs.md:38]:
```suggestion
1. Create a panel with a single query called `edges`.
2. Select your Prometheus data source with metrics from the metrics-generator.
3. Query using the following example:
```

File: docs/sources/tempo/metrics-generator/service_graphs/custom-service-graphs.md
Comment (knylander-grafana) [inline docs/sources/tempo/metrics-generator/service_graphs/custom-service-graphs.md:52]:
```suggestion
1. Use Instant query type.
1. If you encounter issues, switch to the Table data visualization. Refer to the [Node graph panel](/docs/grafana/latest/panels-visualizations/visualizations/node-graph/) documentation for more options and expected data shape.

![Custom service graph panel view](../custom-service-graph-with-query.png)
```

File: docs/sources/tempo/metrics-generator/service_graphs/custom-service-graphs.md
Comment (knylander-grafana) [inline docs/sources/tempo/metrics-generator/service_graphs/custom-service-graphs.md:57]:
```suggestion
All data transformations are done with the Prometheus `label_join` operators. 

Explanation of the query: 
- The first `label_join` creates a new field `id` which is required by the Node graph panel.
- The second and third `label_join` copy the client and server labels to source and target respectively, as these names are also required by Node graph panel.
- `traces_service_graph_request_total` is queried twice with the `OR` operator to get a combination of requests from and to all of the selected services.
```

File: docs/sources/tempo/metrics-generator/service_graphs/custom-service-graphs.md
Comment (knylander-grafana) [inline docs/sources/tempo/metrics-generator/service_graphs/custom-service-graphs.md:63]:
```suggestion
This query gets most of the job done. However, it's limitations can't be compensated for even using the Grafana transform data feature. 
These limitations include:
- unable to add request stats to nodes and edges, such as req/sec and error rates
- unable to add custom icons for nodes

These limitations can be overcome by wrapping the Prometheus query into a REST API, which provides more flexibility with query results data transformations.
You can use [Grafana Infinity data source](/grafana/plugins/yesoreyeram-infinity-datasource/) to visualize data from REST API.
```

**Review by knylander-grafana — CHANGES_REQUESTED** [review]
Thank you for adding such a thorough doc! Excellent explanations. I've provided some feedback to correct grammatical and punctuation errors.

**Review by knylander-grafana — APPROVED** [review]
Thank you for the updates!

**Comment by knylander-grafana** [comment]
I can merge this for you, if you like.

**Comment by ilyamochalov** [comment]
@knylander-grafana  we can merge after someone runs docs project locally to check layouts. I tried `make docs` on my machine but page doesn't load due to some 404s 
Screenshot 2024-06-27 at 08 42 47

Do you know who can build project and check layouts locally?

**Comment by knylander-grafana** [comment]
> @knylander-grafana we can merge after someone runs docs project locally to check layouts. I tried `make docs` on my machine but page doesn't load due to some 404s Screenshot 2024-06-27 at 08 42 47
> 
> Do you know who can build project and check layouts locally?

I can. :)

**Comment by ilyamochalov** [comment]
@knylander-grafana thank you for testing and fixes! If you think the page looks good now, we can merge it.

**Comment by pingping95** [comment]
@ilyamochalov 

It is much more better than original service graph.

really thanks

**Pull request merged**
